apps/system/js/devtools/logshake.js

```diff
@@ -122,52 +156,106 @@
+                                       { operation: error.operation });
         }
       }
 
-      this._notify('logsSaveError', error, moreInfos);
+      return '';
     },
 
-    _notify: function(titleId, body, onclick) {
+    showErrorMessage: function(error, notif) {
+      if (notif) {
+        notif.close();
+      }
+
+      // Do nothing for error string
+      if (typeof error === 'string') {
+        return;
+      }
+
+      if (typeof error !== 'object') {
+        console.warn('Unexpected error type: ' + typeof error);
+        return;
+      }
+
+      // Small heuristic for some frequent unix error cases
+      if ('unixErrno' in error) {
+        var errno = error.unixErrno;
+        debug('errno: ' + errno);
+
+        // Gracefully fallback to a generic error messages if we don't know
+        // this errno code.
+        if (!this.ERRNO_TO_MSG[errno]) {
+          errno = 0;
+        }
+
+        ModalDialog.alert('logsSaveError',
+                          this.ERRNO_TO_MSG[errno], { title: 'ok' });
+      }
+    },
+
+    _notify: function(titleId, body, onclick, dataPayload) {
       var title = navigator.mozL10n.get(titleId) || titleId;
-      var notification =
-        new window.Notification(title, {body: body, tag: 'logshake'});
+      var payload = {
+        body: body,
+        tag: 'logshake',
+        data: {
+          systemMessageTarget: 'logshake',
+          logshakePayload: dataPayload || undefined
+        }
+      };
+      var notification = new Notification(title, payload);
       if (onclick) {
-        notification.onclick = onclick;
+        notification.onclick = onclick.bind(this, notification);
       }
     },
 
-    /**
-     * Stop the component, removing all listeners if necessary
-     */
-    stop: function() {
-      this.stopCaptureLogsListener();
+    handleSystemMessageNotification: function(message) {
+      debug('Received system message: ' + JSON.stringify(message));
+      this.closeSystemMessageNotification(message);
+
+      if (!('logshakePayload' in message.data)) {
+        console.warn('Received logshake system message notification without ' +
+                     'payload, silently discarding.');
+        return;
+      }
+
+      debug('Message payload: ' + message.data.logshakePayload);
+      var payload = message.data.logshakePayload;
+      if ('error' in payload) {
+        this.showErrorMessage(payload.error);
+      } else if ('logFilenames' in payload) {
+        this.triggerShareLogs(payload.logFilenames);
+      } else {
+        console.warn('Unidentified payload: ' + JSON.stringify(payload));
+      }
+
+    },
+
+    closeSystemMessageNotification: function(msg) {
+      Notification.get({ tag: msg.tag }).then(notifs => {
+        notifs.forEach(notif => {
+          if (notif.tag) {
+            // Close notification with the matching tag
+            if (notif.tag === msg.tag) {
+              notif.close && notif.close();
+            }
+          } else {
```

Comment:
This `else` clause is unnecessary and should be removed. First of all, it should indeed have a tag (even legacy logshake notifications had 'logshake' as their tag). But even if it didn't, `Notification.get` with a passed in tag shouldn't return notifications without a tag. And even if it did, we could accidentally close other system notifications that had the same value for body (perhaps empty string?).